### ISSUE TITLE
#[must_use] does not yield warning in pattern match

### ISSUE THREAD

**Issue by jonhoo** [rust-lang/rust#66425]
The following code does not warn that `_f` in `use_foos` is marked as `#[must_use]`:

```rust
#[must_use]
pub struct Foo;
pub enum Foos {
    F(Foo)
}

pub fn use_foos(f: Foos) {
    match f {
        Foos::F(_f) => {}
    }
}
```

If we change `_f` to `f`, we get a warning that `f` is unused, but still no `#[must_use]` warning.

Adding `#[must_use]` to the enum and variant does not make produce the warning either:

```rust
#[must_use]
pub struct Foo;

#[must_use]
pub enum Foos {
    F(#[must_use] Foo)
}

pub fn use_foos(f: Foos) {
    match f {
        Foos::F(_f) => {}
    }
}
```

**Comment by FabianWolff** [comment]
I think this is the intended behavior; see the [reference](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute):
> When used on user-defined composite types, if the expression of an expression statement has that type, then the `unused_must_use` lint is violated.

i.e. you will get the warning if you turn your `match` into an expression statement of type `Foo`:
```rust
    match f {
        Foos::F(f) => f
    };
```
```
warning: unused `Foo` that must be used
  --> src/lib.rs:8:5
   |
8  | /     match f {
9  | |         Foos::F(f) => f
10 | |     };
   | |______^
   |
   = note: `#[warn(unused_must_use)]` on by default
```

**Comment by jonhoo** [comment]
Hmm, this feels like the definition is then wrong — the code doesn't have an expression for `_f` (or make it `f` if you prefer), but it _definitely_ "has" a `Foo` that it's supposed to use lest it violate the lint.
